azure/scope/machinepool.go

```diff
@@ -117,9 +128,77 @@ func NewMachinePoolScope(params MachinePoolScopeParams) (*MachinePoolScope, erro
 	}, nil
 }
 
+func (m *MachinePoolScope) InitMachinePoolCache(ctx context.Context) error {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "azure.MachineScope.InitMachineCache")
+	defer done()
+
+	if m.cache == nil {
+		var err error
+		m.cache = &MachinePoolCache{}
+
+		m.cache.BootstrapData, err = m.GetBootstrapData(ctx)
+		if err != nil {
+			return err
+		}
+
+		m.cache.VMImage, err = m.GetVMImage(ctx)
+		if err != nil {
+			return err
+		}
+
+		skuCache, err := resourceskus.GetCache(m, m.Location())
+		if err != nil {
+			return err
+		}
+
+		m.cache.VMSKU, err = skuCache.Get(ctx, m.AzureMachinePool.Spec.Template.VMSize, resourceskus.VirtualMachines)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get VM SKU %s in compute api", m.AzureMachinePool.Spec.Template.VMSize)
+		}
+
+	}
+
+	return nil
+}
+
 // ScaleSetSpec returns the scale set spec.
-func (m *MachinePoolScope) ScaleSetSpec() azure.ScaleSetSpec {
-	return azure.ScaleSetSpec{
+func (m *MachinePoolScope) ScaleSetSpec(ctx context.Context) azure.ResourceSpecGetter {
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "scope.MachinePoolScope.ScaleSetSpec")
+	defer done()
+
+	vmImage, err := m.GetVMImage(ctx)
+	if err != nil {
+		log.Error(err, "failed to get VM image")
+		// TODO: do we just leave this nil or return an error?
+	}
+
+	bootstrapData, err := m.GetBootstrapData(ctx)
+	if err != nil {
+		log.Error(err, "failed to get bootstrap data")
+		// TODO: do we return early here?
+	}
+
+	maxSurge, err := m.MaxSurge()
+	if err != nil {
+		log.Error(err, "failed to get max surge")
+		// TODO: do we return early here?
+	}
+
+	shouldPatchCustomData := false
+	if m.HasReplicasExternallyManaged(ctx) {
+		shouldPatchCustomData, err := m.HasBootstrapDataChanges(ctx)
```

Comment:
be aware here the assignment doesn't overwrite `shouldPatchCustomData` of the outer scope.